clang/include/clang/Parse/Parser.h

```diff
@@ -2943,6 +2943,9 @@ class Parser : public CodeCompletionHandler {
     return false;
   }
 
+  bool ParseGNUSingleAttribute(ParsedAttributes &Attrs, SourceLocation &EndLoc,
```

Comment:
This could be read as parsing `__attribute__((foo))` as opposed to just `foo` alone. However, I'm not certain I have a better name. `ParseSingleGNUAttributeAfterIntroducer` was the best I could come up with.

If that doesn't make anyone happy, can we switch to `ParseSingleGNUAttribute()` instead?

Reply:
I don't have a preference for any of the 3 names, but a comment explaining WHERE it is in the source would be very helpful.

That said, what makes this GNU specific?  Once we're inside of the braces/bracket/etc, I'd expect there to be only slightly minor differences between the three attribute parsings, so I'd expect them to be a single function with a few special cases (std attributes with an optional namespace, etc).

Reply:
> If that doesn't make anyone happy, can we switch to ParseSingleGNUAttribute() instead?

I think this is my preferred name out of the three.

> That said, what makes this GNU specific? Once we're inside of the braces/bracket/etc, I'd expect there to be only slightly minor differences between the three attribute parsings, so I'd expect them to be a single function with a few special cases (std attributes with an optional namespace, etc).

My main purpose for this refactor is to help aid parsing `__builtin_has_attribute()` (WIP), where the second argument is an attribute. I don't think I'll need to handle non-GNU attributes for that (though I may be wrong, I'll have to check GCC's implementation).

Reply:
Ah, I see.  The WG14 specified one is `__builtin_has_c_attribute`, correct?  

That said, I'd think that we would want the Clang behavior to answer the right thing for ALL of our attributes (including the MSVC ones).  But I then wonder if, since this is a GCC originating builtin, if we should only accept ones with a GCC spelling.

So I guess I'm on the fence here.  I REALLY wish the name was more GCC specific, that way we could just match their behavior, but with the name as it is, I think we have to answer that question for ANY attribute, including declspecs, standard attributes, or clang extensions.

So perhaps that changes this refactor?

Reply:
> Ah, I see. The WG14 specified one is __builtin_has_c_attribute, correct?

No clue. I haven't heard of that builtin before. :-)

> So perhaps that changes this refactor?

I'm not sure how? Let's table discussion about `__builtin_has_attribute` until I get a PR. :-) The GNU documentation for that builtin indicates that it supports only GNU-style attributes. If we need to support more for some reason, then we can refactor other places.

Reply:
>I'm not sure how? 

It would change the purpose of this patch.  If your goal is ot extract this for use in the builtin, and the builtin will work with all attribute names, than this isn't going to be a useful refactor.

>The GNU documentation for that builtin indicates that it supports only GNU-style attributes. If we need to support more for some reason, then we can refactor other places.

My point is: We constantly and consistently 'generalize' the GNU builtins, the GNU documentation, which documents a product that only supports GNU style attributes, claims it only supports GNU style attributes is a bit of a tautology.  

If we take it for clang, we need to generalize it and 'make it work' for all attributes.  BUT I am ok discussing that in the next patch.